File: StringToInteger/string_to_integer.py
class Solution:
    def myAtoi(self, s: str) -> int:
        result = 0
        sign = 1
        i = 0
        n = len(s)

        while i < n and s[i] == " ":
            i += 1

        if i < n and (s[i] == "-" or s[i] == "+"):
            sign = -1 if s[i] == "-" else 1
            i += 1

        while i < n and s[i].isdigit():
            digit = int(s[i])
            if result > (2**31 - 1) // 10 or (
                result == (2**31 - 1) // 10 and digit > 7
            ):
                return 2**31 - 1 if sign == 1 else -(2**31)
            result = result * 10 + digit
            i += 1

        return sign * result

File: 8-StringToInteger/string_to_integer.rs
impl Solution {
    pub fn my_atoi(s: String) -> i32 {
        let mut result = 0;
        let mut sign = 1;
        let mut i = 0;
        let n = s.len();

        // Skip leading whitespace
        while i < n && s.chars().nth(i).unwrap() == ' ' {
            i += 1;
        }

        // Check for sign
        if i < n && (s.chars().nth(i).unwrap() == '-' || s.chars().nth(i).unwrap() == '+') {
            sign = if s.chars().nth(i).unwrap() == '-' { -1 } else { 1 };
            i += 1;
        }

        // Read digits
        while i < n && s.chars().nth(i).unwrap().is_digit(10) {
            let digit = s.chars().nth(i).unwrap() as i32 - '0' as i32;
            // Check for overflow
            if result > (i32::MAX - digit) / 10 {
                return if sign == 1 { i32::MAX } else { i32::MIN };
            }
            result = result * 10 + digit;
            i += 1;
        }

        sign * result
    }
}
